pkg/trace/writer/sender.go

```diff
@@ -1,383 +1,370 @@
 }
 
-func (s *queuableSender) enqueue(p *payload) error {
-	s.mu.Lock()
-	defer s.mu.Unlock()
-	return s.enqueueLocked(p)
+// recordEvent records the occurrence of the given event type t. It additionally
+// passes on the data and augments it with additional information.
+func (s *sender) recordEvent(t eventType, data *eventData) {
+	if recorder := s.cfg.recorder; recorder != nil {
+		data.host = s.cfg.url.Hostname()
+		data.connectionFill = float64(len(s.climit)) / float64(cap(s.climit))
+		data.queueFill = float64(len(s.queue)) / float64(cap(s.queue))
+		recorder.recordEvent(t, data)
+	}
 }
 
-func (s *queuableSender) enqueueLocked(payload *payload) error {
-	// Start by discarding payloads that are too old, freeing up memory
-	s.discardOldPayloads()
+// userAgent is the computed user agent we'll use when communicating with Datadog var
+var userAgent = fmt.Sprintf("Datadog Trace Agent/%s/%s", info.Version, info.GitCommit)
 
-	for s.conf.MaxQueuedPayloads > 0 && s.queuedPayloads.Len() >= s.conf.MaxQueuedPayloads {
-		log.Debugf("Dropping existing payload because max queued payloads reached: %d", s.conf.MaxQueuedPayloads)
-		if _, err := s.dropOldestPayload("max queued payloads reached"); err != nil {
-			panic(fmt.Errorf("unable to respect max queued payloads value of %d", s.conf.MaxQueuedPayloads))
-		}
-	}
+// retriableError is an error returned by the server which may be retried at a later time.
+type retriableError struct{ err error }
 
-	newPayloadSize := int64(len(payload.bytes))
+// Error implements error.
+func (e retriableError) Error() string { return e.err.Error() }
 
-	if s.conf.MaxQueuedBytes > 0 && newPayloadSize > s.conf.MaxQueuedBytes {
-		log.Debugf("Payload bigger than max size: size=%d, max size=%d", newPayloadSize, s.conf.MaxQueuedBytes)
-		return fmt.Errorf("unable to queue payload bigger than max size: payload size=%d, max size=%d",
-			newPayloadSize, s.conf.MaxQueuedBytes)
-	}
+const (
+	headerAPIKey    = "DD-Api-Key"
+	headerUserAgent = "User-Agent"
+)
 
-	for s.conf.MaxQueuedBytes > 0 && s.currentQueuedSize+newPayloadSize > s.conf.MaxQueuedBytes {
-		if _, err := s.dropOldestPayload("max queued bytes reached"); err != nil {
-			// Should never happen because we know we can fit it in
-			panic(fmt.Errorf("unable to find space for queueing payload of size %d: %v", newPayloadSize, err))
+func (s *sender) do(req *http.Request) error {
+	req.Header.Set(headerAPIKey, s.cfg.apiKey)
+	req.Header.Set(headerUserAgent, userAgent)
+	resp, err := s.cfg.client.Do(req)
+	if err != nil {
+		// request errors are either redirect errors, url errors
+		return &retriableError{err}
+	}
+	if resp.StatusCode/100 == 5 {
+		// 5xx errors can be retried
+		return &retriableError{
+			fmt.Errorf("server responded with %q", resp.Status),
 		}
 	}
-
-	log.Tracef("Queuing new payload: %v", payload)
-	s.queuedPayloads.PushBack(payload)
-	s.currentQueuedSize += newPayloadSize
-
+	if resp.StatusCode/100 != 2 {
+		// non-2xx errors are failures
```

Comment:
```suggestion
		// status codes that are neither 2xx nor 5xx are considered non-retriable failures
```

Making it more explicit.

Reply:
Done.